File: CS/Form1.cs
﻿#region usings
using DevExpress.DataAccess.Json;
using DevExpress.XtraReports.UI;
#endregion
using System;
using System.Net;
using System.Windows.Forms;


namespace Create_a_Report_Bound_to_JsonDataSource
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            // Enables security protocol versions to access the JSON web data source.
            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls
                | SecurityProtocolType.Tls11
                | SecurityProtocolType.Tls12
                | SecurityProtocolType.Ssl3;

            XtraReport report = CreateReport();
            ReportDesignTool designTool = new ReportDesignTool(report);
            designTool.ShowRibbonDesignerDialog();
            Application.Exit();
        }
        #region CreateReport_start
        private XtraReport CreateReport() {
            XtraReport report = new XtraReport() {
                Bands = {
                    new DetailBand() {
                        Controls = {
                            new XRLabel() {
                                ExpressionBindings = {
                                    new ExpressionBinding("BeforePrint", "Text", "[CompanyName]")
                                },
                                WidthF = 300
                            }
                        },
                        HeightF = 50
                    }
                },
                DataSource = CreateDataSourceFromWeb(),
                #endregion
                //DataSource = CreateDataSourceFromFile(),
                //DataSource = CreateDataSourceFromText(),
                #region CreateReport_end
                DataMember = "Customers"
            };
            return report;
        }
        #endregion
        #region CreateDataSourceFromWeb_start
        public static JsonDataSource CreateDataSourceFromWeb() {
            var jsonDataSource = new JsonDataSource();
            // Specify the endpoint.
            jsonDataSource.JsonSource = new UriJsonSource(
                new Uri("https://raw.githubusercontent.com/DevExpress-Examples/DataSources/master/JSON/customers.json"));
            #endregion
            #region CreateDataSourceFromWeb_schema
            var root = new JsonSchemaNode();
            root.NodeType = JsonNodeType.Object;

            var customers = new JsonSchemaNode() {NodeType=JsonNodeType.Array, Name="Customers", Selected=true };
            customers.AddChildren(new[] {
                new JsonSchemaNode(new JsonNode("CustomerID", true,
                JsonNodeType.Property, typeof(string))) 
                { 
                DisplayName = "Customer ID" },
                new JsonSchemaNode() {
                    Name =  "CompanyName",
                    Selected = true,
                    NodeType = JsonNodeType.Property,
                    Type = typeof(string)
                },
                new JsonSchemaNode(new JsonNode("ContactTitle", true, JsonNodeType.Property, typeof(string))),
                new JsonSchemaNode(new JsonNode("Address", false, JsonNodeType.Property, typeof(string)))
            });

            root.AddChildren(customers);
            jsonDataSource.Schema = root;
            #endregion
            #region CreateDataSourceFromWeb_end
            // The schema is built, you do not have to call the Fill method to populate the Field List.
			// The Designer calls the Fill method automatically when a document is generated for preview.
            //jsonDataSource.Fill();
            return jsonDataSource;
        }
        #endregion
        #region CreateDataSourceFromFile
        public static JsonDataSource CreateDataSourceFromFile() {
            var jsonDataSource = new JsonDataSource();
            // Specify the JSON file name.
            Uri fileUri = new Uri("customers.json", UriKind.RelativeOrAbsolute);
            jsonDataSource.JsonSource = new UriJsonSource(fileUri);
            // Populate the data source with data.
            jsonDataSource.Fill();
            return jsonDataSource;
        }
        #endregion
        #region CreateDataSourceFromText
        public static JsonDataSource CreateDataSourceFromText() {
            var jsonDataSource = new JsonDataSource();

            // Specify a string with JSON data.
            string json = "{\"Customers\":[{\"Id\":\"ALFKI\",\"CompanyName\":\"Alfreds Futterkiste\"," +
                "\"ContactName\":\"Maria Anders\",\"ContactTitle\":\"Sales Representative\"," +
                "\"Address\":\"Obere Str. 57\",\"City\":\"Berlin\",\"PostalCode\":\"12209\"," +
                "\"Country\":\"Germany\",\"Phone\":\"030-0074321\",\"Fax\":\"030-0076545\"}]," +
                "\"ResponseStatus\":{}}";

            // Specify the object that retrieves JSON data.
            jsonDataSource.JsonSource = new CustomJsonSource(json);
            // Populate the data source with data.
            jsonDataSource.Fill();
            return jsonDataSource;
        }
        #endregion
    }
}

File: VB/Form1.vb
#Region "usings"
Imports System.Net
Imports DevExpress.DataAccess.Json
Imports DevExpress.XtraReports.UI
#End Region


Namespace Create_a_Report_Bound_to_JsonDataSource
    Partial Public Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load

            'Enables security protocol versions to access the JSON web data source.
            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls Or
                SecurityProtocolType.Tls11 Or
                SecurityProtocolType.Tls12 Or
                SecurityProtocolType.Ssl3

            Dim report As XtraReport = CreateReport()
            Dim designTool As New ReportDesignTool(report)
            designTool.ShowRibbonDesignerDialog()
            Application.Exit()
        End Sub
#Region "CreateReport_start"
        Private Function CreateReport() As XtraReport
            Dim report As New XtraReport()
            Dim DetailBand As New DetailBand()
            DetailBand.HeightF = 50

            Dim XRLabel As New XRLabel()
            XRLabel.WidthF = 300
            XRLabel.ExpressionBindings.Add(New ExpressionBinding("BeforePrint", "Text", "[CompanyName]"))

            DetailBand.Controls.Add(XRLabel)
            report.Bands.Add(DetailBand)
			
            report.DataSource = CreateDataSourceFromWeb()
            'report.DataSource = CreateDataSourceFromFile();
            'report.DataSource = CreateDataSourceFromText();
            report.DataMember = "Customers"
            Return report
        End Function
#End Region
#Region "CreateDataSourceFromWeb"
        Public Shared Function CreateDataSourceFromWeb() As JsonDataSource
            Dim jsonDataSource = New JsonDataSource()
            ' Specify the endpoint.
            jsonDataSource.JsonSource = New UriJsonSource(
                New Uri("https://raw.githubusercontent.com/DevExpress-Examples/DataSources/master/JSON/customers.json"))
            Dim root = New JsonSchemaNode()
            root.NodeType = JsonNodeType.Object

            Dim customers = New JsonSchemaNode() With {.NodeType = JsonNodeType.Array, .Name = "Customers", .Selected = True}
            customers.AddChildren({
                New JsonSchemaNode(New JsonNode("CustomerID", True, JsonNodeType.Property,
                                                GetType(String))) With {.DisplayName = "Customer ID"},
                New JsonSchemaNode() With {.Name = "CompanyName", .Selected = True,
                .NodeType = JsonNodeType.Property, .Type = GetType(String)},
                New JsonSchemaNode(New JsonNode("ContactTitle", True, JsonNodeType.Property,
                                                GetType(String))),
                New JsonSchemaNode(New JsonNode("Address", False, JsonNodeType.Property,
                                                GetType(String)))
            })

            root.AddChildren(customers)
            jsonDataSource.Schema = root
            ' Populate the data source with data.
            jsonDataSource.Fill()
            Return jsonDataSource
        End Function
#End Region
#Region "CreateDataSourceFromFile"
        Public Shared Function CreateDataSourceFromFile() As JsonDataSource
            Dim jsonDataSource = New JsonDataSource()
            ' Specify the JSON file name.
            Dim fileUri As New Uri("customers.json", UriKind.RelativeOrAbsolute)
            jsonDataSource.JsonSource = New UriJsonSource(fileUri)
            ' Populate the data source with data.
            jsonDataSource.Fill()
            Return jsonDataSource
        End Function
#End Region
#Region "CreateDataSourceFromText"
        Public Shared Function CreateDataSourceFromText() As JsonDataSource
            Dim jsonDataSource = New JsonDataSource()

            ' Specify a string with JSON data.
            Dim json As String = "{""Customers"":[{""Id"":""ALFKI"",""CompanyName"":""Alfreds Futterkiste""," +
                """ContactName"":""Maria Anders"",""ContactTitle"":""Sales Representative""," +
            """Address"":""Obere Str. 57"",""City"":""Berlin"",""PostalCode"":""12209""," +
            """Country"":""Germany"",""Phone"":""030-0074321""," +
            """Fax"":""030-0076545""}],""ResponseStatus"":{}}"

            ' Specify the object that retrieves JSON data.
            jsonDataSource.JsonSource = New CustomJsonSource(json)
            ' Populate the data source with data.
            jsonDataSource.Fill()
            Return jsonDataSource
        End Function
#End Region
    End Class
End Namespace
